src/server/util/insertData.js

```diff
@@ -20,26 +20,51 @@ const fs = require('fs').promises;
 async function insertUnits(unitsToInsert, update = false, conn) {
 	await Promise.all(unitsToInsert.map(
 		async unitData => {
-			const dbUnit = await Unit.getByName(unitData[0], conn);
-			if (dbUnit === null) {
-				// The unit does not exist so add it.
-				await new Unit(undefined, unitData[0], unitData[1], unitData[2], unitData[3],
-					unitData[4], null, unitData[5], unitData[6], unitData[7], unitData[8]).insert(conn);
-			} else if (update) {
-				// Asked to update so will. Does not bother to check if no changes.
-				dbUnit.name = unitData[0];
-				dbUnit.identifier = unitData[1];
-				dbUnit.unitRepresent = unitData[2];
-				dbUnit.secInRate = unitData[3];
-				dbUnit.typeOfUnit = unitData[4];
-				dbUnit.suffix = unitData[5];
-				dbUnit.displayable = unitData[6];
-				dbUnit.preferredDisplay = unitData[7];
-				dbUnit.note = unitData[8];
-				// Should update even though exists.
-				await dbUnit.update(conn);
+			let ok = true;
+			for (let i = 0; i < unitsToInsert.length; ++i) {
+				// Meter key/value pairs for the current meter.
+				const uData = unitsToInsert[i];
+		
+				// Check that needed keys are there.
+				const requiredKeys = ['name', 'unitRepresent', 'typeOfUnit', 'displayable', 'preferredDisplay'];
+				ok = true;
+				requiredKeys.forEach(key => {
+					if ((key == 'name') && (typeof uData[key] !== 'string')){
+						console.log(`********key "${key}" is required but missing so unit number ${i} not processed`);
+						ok = false;
+					}
+					if (((key == 'unitRepresent') || (key == 'typeOfUnit') || (key == 'displayable')) && (typeof uData[key] !== 'string')){
+						console.log(`********key "${key}" is required but missing so conversion number ${i} not processed`);
+						ok = false;
+					}
+					if ((key == 'preferredDisplay') && (typeof uData[key] !== 'boolean')){
+						console.log(`********key "${key}" is required but missing so unit number ${i} not processed`);
+						ok = false;
+					}
+				})
+			}
+			if (ok) {
+				const dbUnit = await Unit.getByName(unitData.name, conn);
+				if (dbUnit === null) {
+					// The unit does not exist so add it.
+					await new Unit(undefined, unitData.name, unitData.identifier, unitData.unitRepresent, unitData.secInRate,
+						unitData.typeOfUnit, null, unitData.suffix, unitData.displayable, unitData.preferredDisplay, unitData.note).insert(conn);
+				} else if (update) {
+					// Asked to update so will. Does not bother to check if no changes.
+					dbUnit.name = unitData.name;
+					dbUnit.identifier = unitData.identifier;
+					dbUnit.unitRepresent = unitData.unitRepresent;
+					dbUnit.secInRate = unitData.secInRate;
+					dbUnit.typeOfUnit = unitData.typeOfUnit;
+					dbUnit.suffix = unitData.suffix;
+					dbUnit.displayable = unitData.displayable;
+					dbUnit.preferredDisplay = unitData.preferredDisplay;
+					dbUnit.note = unitData.note;
+					// Should update even though exists.
+					await dbUnit.update(conn);
+					// Otherwise do not update.
```

Comment:
I'm unclear on why this comment moved. The line above just did the update. I think it was correct where it was.